File: web/sprite-motion.js
(function (root) {
  'use strict';
  const DIRECTIONS = ['right', 'left', 'up', 'down', 'up_right', 'up_left', 'down_right', 'down_left'];
  const TRAVEL_LIMITS = {right: 1, left: 1, up: .5, down: .5,
    up_right: 1, up_left: 1, down_right: 1, down_left: 1};
  const MAX_DELTA = 1 / 30;
  const quintic = t => t * t * t * (t * (6 * t - 15) + 10);
  function inverseQuintic(value) {
    if (value <= 0 || value >= 1) return Math.max(0, Math.min(1, value));
    let low = 0, high = 1;
    for (let i = 0; i < 30; i++) {
      const middle = (low + high) / 2;
      if (quintic(middle) < value) low = middle; else high = middle;
    }
    return (low + high) / 2;
  }

  class SpriteMotion {
    constructor({count = 24, random = Math.random, playing = true} = {}) {
      if (!Number.isInteger(count) || count < 2) throw new Error('At least two sprite frames are required.');
      this.count = count;
      this.random = random;
      this.playing = playing;
      this.direction = 'right';
      this.index = 0;
      this.target = count - 1;
      this.phase = 'center';
      this.progress = 0;
      this.hold = .7;
      this.duration = 1.8;
      this.speed = 1;
      this.eased = true;
      this.auto = true;
      this.cycling = false;
      this.cycleIndex = 0;
      this.queue = [];
      this.inspecting = false;
      this.blinks = true;
      this.blinkLevel = 0;
      this.blinkPhase = 'idle';
      this.blinkElapsed = 0;
      this.blinkWait = this.range(3, 7);
      this.blinkQueued = false;
      this.doublePending = false;
    }
    range(low, high) { return low + this.random() * (high - low); }
    setPlaying(value) { this.playing = Boolean(value); if (value) this.inspecting = false; }
    setDuration(value) {
      if (!Number.isFinite(value) || value <= 0) throw new Error('Turn duration must be positive.');
      this.duration = value;
    }
    setSpeed(value) {
      if (!Number.isFinite(value) || value <= 0) throw new Error('Playback speed must be positive.');
      this.speed = value;
    }
    setEased(value) { this.eased = Boolean(value); }
    setAuto(value) { this.auto = Boolean(value); if (value) this.cycling = false; }
    setCycle(value) {
      this.cycling = Boolean(value);
      if (value) { this.auto = false; this.cycleIndex = 0; }
    }
    request(direction) {
      if (!DIRECTIONS.includes(direction)) throw new Error(`Unknown sprite direction: ${direction}`);
      this.queue.push(direction);
    }
    inspect(index) {
      this.setPlaying(false);
      this.inspecting = true;
      this.index = Math.max(0, Math.min(this.count - 1, Math.round(index)));
      // Inspection is an explicit frame selection, never a live track switch.
      // Resume from that exact pose, safely retracing it to the shared center.
      this.phase = this.index ? 'return' : 'center';
      this.target = Math.max(1, this.index);
      this.progress = 0;
    }
    setBlinks(value) {
      this.blinks = Boolean(value);
      if (!value) this.doublePending = false;
      // An active blink always opens normally, including when disabled.
    }
    requestBlink() { this.blinkQueued = true; }
    startBlink() {
      this.blinkPhase = 'closing';
      this.blinkElapsed = 0;
      this.blinkQueued = false;
    }
    updateBlink(dt) {
      if (this.blinkPhase === 'idle') {
        if (this.blinks || this.doublePending) this.blinkWait -= dt;
        if (this.blinkQueued || ((this.blinks || this.doublePending) && this.blinkWait <= 0)) {
          const second = this.doublePending;
          this.doublePending = !second && this.blinks && this.random() < .15;
          this.startBlink();
          return true;
        }
        return false;
      }
      this.blinkElapsed += dt;
      const interval = this.blinkPhase === 'closing' ? .09 / 4
        : this.blinkPhase === 'closed' ? .04 : .14 / 4;
      if (this.blinkElapsed < interval) return true;
      this.blinkElapsed = Math.min(this.blinkElapsed - interval, interval);
      if (this.blinkPhase === 'closing') {
        this.blinkLevel++;
        if (this.blinkLevel === 4) { this.blinkPhase = 'closed'; this.blinkElapsed = 0; }
      } else if (this.blinkPhase === 'closed') {
        this.blinkPhase = 'opening';
        this.blinkElapsed = 0;
      } else {
        this.blinkLevel--;
        if (this.blinkLevel === 0) {
          this.blinkPhase = 'idle';
          this.blinkWait = this.doublePending ? this.range(.16, .28) : this.range(3, 7);
        }
      }
      return true;
    }
    startTurn() {
      let direction, fullDepth = true;
      if (this.queue.length) direction = this.queue.shift();
      else if (this.cycling) direction = DIRECTIONS[this.cycleIndex++ % DIRECTIONS.length];
      else if (this.auto) {
        const choices = DIRECTIONS.filter(value => value !== this.direction);
        direction = choices[Math.min(choices.length - 1, Math.floor(this.random() * choices.length))];
        fullDepth = false;
      } else return;
      this.direction = direction;
      const limit = Math.max(1, Math.floor((this.count - 1) * TRAVEL_LIMITS[direction]));
      this.target = fullDepth ? limit : Math.round(this.range(.55, 1) * limit);
      this.phase = 'out';
      this.progress = 0;
    }
    edgeDuration() {
      const travelled = this.phase === 'out' ? this.index : this.target - this.index;
      const before = travelled / this.target, after = (travelled + 1) / this.target;
      const turnDuration = this.duration * Math.sqrt(this.target / (this.count - 1));
      return turnDuration * (this.eased ? inverseQuintic(after) - inverseQuintic(before) : after - before);
    }
    sample(crossfade = false) {
      if (!crossfade || this.inspecting || (this.phase !== 'out' && this.phase !== 'return')) {
        return {from: this.index, to: this.index, mix: 0};
      }
      return {
        from: this.index,
        to: this.phase === 'out' ? Math.min(this.target, this.index + 1) : Math.max(0, this.index - 1),
        mix: Math.max(0, Math.min(1, this.progress)),
      };
    }
    update(deltaSeconds) {
      if (!this.playing || !Number.isFinite(deltaSeconds) || deltaSeconds <= 0) return this;
      // Drop wall-clock stalls, rather than accumulating a catch-up timeline.
      const dt = Math.min(deltaSeconds, MAX_DELTA) * this.speed;
      this.updateBlink(Math.min(dt, MAX_DELTA));
      if (this.phase === 'center' || this.phase === 'endpoint') {
        this.progress += dt / this.hold;
        if (this.progress < 1) return this;
        this.progress = 0;
        if (this.phase === 'center') this.startTurn();
        else this.phase = 'return';
        return this;
      }
      const interval = this.edgeDuration();
      this.progress += dt / interval;
      if (this.progress < 1) return this;
      const remainder = (this.progress - 1) * interval;
      this.progress = 0;
      this.index += this.phase === 'out' ? 1 : -1;
      if (this.index === this.target && this.phase === 'out') {
        this.phase = 'endpoint';
        this.hold = this.cycling ? .6 : this.range(.35, 1.4);
      } else if (this.index === 0 && this.phase === 'return') {
        this.phase = 'center';
        this.hold = this.cycling ? .6 : this.range(.35, 1.6);
      } else {
        // Preserve fractional frame time without accumulating a catch-up backlog.
        this.progress = Math.min(remainder / this.edgeDuration(), 1);
      }
      return this;
    }
  }
  const api = {SpriteMotion, DIRECTIONS, quintic, inverseQuintic};
  if (typeof module !== 'undefined' && module.exports) module.exports = api;
  else root.SpritePlayer = api;
})(typeof globalThis !== 'undefined' ? globalThis : this);

File: firmware/Copilot/src/SpriteMotion.cpp
#include "SpriteMotion.h"
#include <algorithm>
#include <cmath>

namespace copilot {
namespace {
constexpr double kMaxDelta = 1.0 / 30;
constexpr const char* kInvalidCount = "At least two sprite frames are required.";

double quintic(double t) {
  return t * t * t * (t * (6 * t - 15) + 10);
}

double inverseQuintic(double value) {
  if (value <= 0 || value >= 1) return std::max(0.0, std::min(1.0, value));
  double low = 0, high = 1;
  for (int i = 0; i < 30; ++i) {
    const double middle = (low + high) / 2;
    if (quintic(middle) < value) low = middle;
    else high = middle;
  }
  return (low + high) / 2;
}
}

SpriteMotion::SpriteMotion(uint32_t seed, uint8_t frameCount)
    : rng_(seed ? seed : 0x6d2b79f5u), count_(frameCount),
      target_(frameCount >= 2 ? frameCount - 1 : 0), blinkWait_(range(3, 7)) {
  if (!valid()) error_ = kInvalidCount;
}

double SpriteMotion::random() {
  rng_ ^= rng_ << 13;
  rng_ ^= rng_ >> 17;
  rng_ ^= rng_ << 5;
  return static_cast<double>(rng_) / 4294967296.0;
}

double SpriteMotion::range(double low, double high) {
  return low + random() * (high - low);
}

bool SpriteMotion::fail(const char* message) {
  error_ = message;
  return false;
}

void SpriteMotion::clearError() {
  error_ = valid() ? nullptr : kInvalidCount;
}

void SpriteMotion::setAutomatic(bool value) {
  automatic_ = value;
  if (value) cycling_ = false;
}

void SpriteMotion::setCycle(bool value) {
  cycling_ = value;
  if (value) {
    automatic_ = false;
    cycleIndex_ = 0;
  }
}

void SpriteMotion::setBlinks(bool value) {
  blinks_ = value;
  if (!value) doublePending_ = false;
}

bool SpriteMotion::setDuration(double value) {
  if (!std::isfinite(value) || value <= 0) return fail("Turn duration must be positive.");
  duration_ = value;
  clearError();
  return valid();
}

bool SpriteMotion::setSpeed(double value) {
  if (!std::isfinite(value) || value <= 0) return fail("Playback speed must be positive.");
  speed_ = value;
  clearError();
  return valid();
}

bool SpriteMotion::request(int direction, double depth) {
  if (!valid()) return fail(kInvalidCount);
  if (direction < 0 || direction >= 8) return fail("Unknown sprite direction.");
  if (!std::isfinite(depth) || depth <= 0 || depth > 1)
    return fail("Sprite direction depth must be between zero and one.");
  if (queueSize_ == kQueueCapacity) return fail("Sprite direction queue is full.");
  queue_[(queueHead_ + queueSize_) % kQueueCapacity] = {
      static_cast<uint8_t>(direction), depth};
  ++queueSize_;
  clearError();
  return true;
}

bool SpriteMotion::returnToCenter(double duration) {
  if (!setDuration(duration)) return false;
  automatic_ = cycling_ = false;
  queueHead_ = queueSize_ = 0;
  if (phase_ == Phase::Return || (phase_ == Phase::Center && !pose_.index)) return true;
  target_ = std::max<uint8_t>(1, pose_.index);
  phase_ = pose_.index ? Phase::Return : Phase::Center;
  progress_ = 0;
  return true;
}

void SpriteMotion::startTurn() {
  uint8_t direction;
  double depth = 1;
  if (queueSize_) {
    const Request request = queue_[queueHead_];
    direction = request.direction;
    depth = request.depth;
    queueHead_ = (queueHead_ + 1) % kQueueCapacity;
    --queueSize_;
  } else if (cycling_) {
    direction = cycleIndex_;
    cycleIndex_ = (cycleIndex_ + 1) % 8;
  } else if (automatic_) {
    direction = static_cast<uint8_t>(std::floor(random() * 7));
    if (direction >= pose_.direction) ++direction;
    depth = range(.55, 1);
  } else {
    return;
  }
  pose_.direction = direction;
  const int limit = std::max(1, (direction == 2 || direction == 3)
      ? (count_ - 1) / 2 : count_ - 1);
  target_ = static_cast<uint8_t>(std::max(1.0, std::round(depth * limit)));
  phase_ = Phase::Out;
  progress_ = 0;
}

double SpriteMotion::edgeDuration() {
  const int travelled = phase_ == Phase::Out ? pose_.index : target_ - pose_.index;
  if (edgeCache_.target == target_ && edgeCache_.travelled == travelled
      && edgeCache_.eased == eased_ && edgeCache_.duration == duration_) {
    return edgeCache_.interval;
  }
  const double before = static_cast<double>(travelled) / target_;
  const double after = static_cast<double>(travelled + 1) / target_;
  const double turnDuration = duration_ * std::sqrt(static_cast<double>(target_) / (count_ - 1));
  const double interval = turnDuration
      * (eased_ ? inverseQuintic(after) - inverseQuintic(before) : after - before);
  edgeCache_ = {duration_, interval, target_, static_cast<uint8_t>(travelled), eased_};
  return interval;
}

void SpriteMotion::updateBlink(double dt) {
  if (blinkPhase_ == BlinkPhase::Idle) {
    if (blinks_ || doublePending_) blinkWait_ -= dt;
    if (blinkQueued_ || ((blinks_ || doublePending_) && blinkWait_ <= 0)) {
      const bool second = doublePending_;
      doublePending_ = !second && blinks_ && random() < .15;
      blinkPhase_ = BlinkPhase::Closing;
      blinkElapsed_ = 0;
      blinkQueued_ = false;
    }
    return;
  }
  blinkElapsed_ += dt;
  const double interval = blinkPhase_ == BlinkPhase::Closing ? .09 / 4
      : blinkPhase_ == BlinkPhase::Closed ? .04 : .14 / 4;
  if (blinkElapsed_ < interval) return;
  blinkElapsed_ = std::min(blinkElapsed_ - interval, interval);
  if (blinkPhase_ == BlinkPhase::Closing) {
    ++pose_.blinkLevel;
    if (pose_.blinkLevel == 4) {
      blinkPhase_ = BlinkPhase::Closed;
      blinkElapsed_ = 0;
    }
  } else if (blinkPhase_ == BlinkPhase::Closed) {
    blinkPhase_ = BlinkPhase::Opening;
    blinkElapsed_ = 0;
  } else {
    --pose_.blinkLevel;
    if (pose_.blinkLevel == 0) {
      blinkPhase_ = BlinkPhase::Idle;
      blinkWait_ = doublePending_ ? range(.16, .28) : range(3, 7);
    }
  }
}

void SpriteMotion::update(double deltaSeconds) {
  if (!valid() || !playing_ || !std::isfinite(deltaSeconds) || deltaSeconds <= 0) return;
  const double dt = std::min(deltaSeconds, kMaxDelta) * speed_;
  // Blinking is independent: it must never add a dwell to the head timeline.
  updateBlink(std::min(dt, kMaxDelta));
  if (phase_ == Phase::Center || phase_ == Phase::Endpoint) {
    progress_ += dt / hold_;
    if (progress_ < 1) return;
    progress_ = 0;
    if (phase_ == Phase::Center) startTurn();
    else phase_ = Phase::Return;
    return;
  }
  const double interval = edgeDuration();
  progress_ += dt / interval;
  if (progress_ < 1) return;
  const double remainder = interval > 0 ? (progress_ - 1) * interval : 0;
  progress_ = 0;
  if (phase_ == Phase::Out) ++pose_.index;
  else --pose_.index;
  if (pose_.index == target_ && phase_ == Phase::Out) {
    phase_ = Phase::Endpoint;
    hold_ = cycling_ ? .6 : range(.35, 1.4);
  } else if (pose_.index == 0 && phase_ == Phase::Return) {
    phase_ = Phase::Center;
    hold_ = cycling_ ? .6 : range(.35, 1.6);
  } else {
    // Keep ordinary fractional timing, but never queue more than one edge.
    const double next = edgeDuration();
    progress_ = next > 0 ? std::min(remainder / next, 1.0) : 1;
  }
}
}
